### DISCUSSION TITLE
StreamingResponse with async generator sometimes causes socket.send() exception indefinitely

### DISCUSSION THREAD

**Discussion by MichadeGroot** [fastapi/fastapi#6260] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio

import uvicorn
from fastapi import FastAPI
from fastapi.responses import StreamingResponse

app = FastAPI()


def sync_streamer():
    try:
        while True:
            yield b"--boundary\r\nContent-Type: text/plain\r\nContent-Length: 1\r\n\r\n1\r\n"
    except asyncio.CancelledError:
        print("caught cancelled error")


async def async_streamer():
    try:
        while True:
            yield b"--boundary\r\nContent-Type: text/plain\r\nContent-Length: 1\r\n\r\n1\r\n"
    except asyncio.CancelledError:
        print("caught cancelled error")


@app.get("/sync")
async def sync_endpoint():
    return StreamingResponse(
        sync_streamer(),
        media_type="multipart/x-mixed-replace; boundary=boundary",
    )


@app.get("/async")
async def async_endpoint():
    return StreamingResponse(
        async_streamer(),
        media_type="multipart/x-mixed-replace; boundary=boundary",
    )
```

### Description

When I try to send data with a StreamingResponse from an infinite generator I sometimes get an error when the generator is asynchronous. Sometimes it works fine, but sometimes I get an infinite stream of errors: `socket.send() raised exception.` on the server side. 
If I instead use a synchronous generator this problem does not appear at all.

I've based my sample code on #4146 and the `CancelledError` does get caught sometimes, but sometimes it doesn't. Right now I can avoid this issue by making the generator synchronous, but it would be nice to have the option to use async generators as well.

It could be that this is a Starlette issue but I am really not sure.

### Operating System

Linux

### FastAPI Version

0.79.0

### Python Version

3.8.10

**Comment by iudeen** [comment]
Even `KeyboardInterrupt` is not getting caught

**Comment by Kludex** [comment]
What's the exception?

> It could be that this is a Starlette issue but I am really not sure.

If you can reproduce with a pure Starlette application, then I'd invite you to create a discussion on Starlette's side. 🙏

**Comment by iudeen** [comment]
> What's the exception?
> 
> 
> 
> > It could be that this is a Starlette issue but I am really not sure.
> 
> 
> 
> If you can reproduce with a pure Starlette application, then I'd invite you to create a discussion on Starlette's side. 🙏

No exceptions at all, it's stuck endlessly. 

I'll do that!

**Comment by iudeen** [comment, marked as answer]
@MichadeGroot as per @Kludex there is no `await` in your async generators, so all other tasks in task groups are waiting endlesssly for it to complete.

The solution for it is here https://github.com/encode/starlette/discussions/1776#discussioncomment-3207518

**Comment by MichadeGroot** [comment]
Thank you @Kludex and @iudeen for figuring out the problem!
